File: 4.8-bc/go/stage-2-feature-first/app/commerce/handler.go
package commerce

import (
	"net/http"

	"github.com/go-chi/chi/v5"
	"github.com/google/uuid"

	"github.com/genkovich/claude-course-demos/4.8-bc/go/stage-2-feature-first/shared/httputil"
)

type Handler struct {
	svc *Service
}

func NewHandler(svc *Service) *Handler {
	return &Handler{svc: svc}
}

func (h *Handler) Register(r chi.Router) {
	r.Post("/orders", h.place)
}

type placeReq struct {
	UserID     string `json:"user_id"`
	TotalCents int64  `json:"total_cents"`
}

func (h *Handler) place(w http.ResponseWriter, r *http.Request) {
	var req placeReq
	if err := httputil.DecodeBody(r, &req); err != nil {
		httputil.WriteJSON(w, http.StatusBadRequest, map[string]string{"error": "invalid_request"})
		return
	}
	uid, err := uuid.Parse(req.UserID)
	if err != nil {
		httputil.WriteJSON(w, http.StatusBadRequest, map[string]string{"error": "invalid_user_id"})
		return
	}
	o, err := h.svc.Place(r.Context(), uid, req.TotalCents)
	if err != nil {
		httputil.WriteError(w, err)
		return
	}
	httputil.WriteJSON(w, http.StatusCreated, map[string]string{"order_id": o.ID.String(), "status": o.Status})
}

File: 4.8-bc/py/stage-3-hexagonal/app/commerce/infra/http/handler.py
"""Commerce HTTP handler — POST /orders."""
from __future__ import annotations

from uuid import UUID

from fastapi import APIRouter, HTTPException, status
from pydantic import BaseModel

from app.commerce.app.service import Service
from app.commerce.domain.order import InvalidOrderError


class PlaceReq(BaseModel):
    user_id: str
    total_cents: int


class PlaceResp(BaseModel):
    order_id: str
    status: str


def build_router(svc: Service) -> APIRouter:
    router = APIRouter(tags=["commerce"])

    @router.post(
        "/orders",
        status_code=status.HTTP_201_CREATED,
        response_model=PlaceResp,
    )
    async def place_order(req: PlaceReq) -> PlaceResp:
        try:
            uid = UUID(req.user_id)
        except (ValueError, TypeError) as exc:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST, detail="invalid_user_id"
            ) from exc
        try:
            o = await svc.place(uid, req.total_cents)
        except InvalidOrderError as exc:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail={
                    "error": "commerce.invalid_order",
                    "message": "total_cents must be >= 0",
                },
            ) from exc
        return PlaceResp(order_id=str(o.id), status=o.status)

    return router
